static/js/compose_ui.js

```diff
@@ -372,23 +405,199 @@ export function format_text(textarea, type) {
+        const description = selected_text.split("](")[0].replace("[", "");
+        let url = selected_text.split("](")[1].replace(")", "");
+        url = url === "url" ? "" : " " + url;
+        text = text.slice(0, range.start) + description + url + text.slice(range.end, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start - 1, range.start - 1);
+        return;
+    }
+
+    // Otherwise, we don't have code syntax, so we add it.
+    wrapSelection(field, link_syntax_start, link_syntax_end);
+
+    // Change selected text to `url` part of the syntax.
+    // If <text> marks the selected region, we're mapping:
+    // <text> => [text](<url>).
+    const new_start = range.end + "[](".length;
+    const new_end = new_start + "url".length;
+    field.setSelectionRange(new_start, new_end);
+}
+
+function add_spoiler_formatting(field, range, text, selected_text) {
+    let spoiler_syntax_start = "```spoiler \n";
+    let spoiler_syntax_end = "\n```";
+
+    const is_curser_at_header_position = () =>
+        range.length === 0 && text.slice(range.start - 11, range.start) === "```spoiler ";
+
+    const is_content_selected = () =>
+        text.slice(range.end, range.end + spoiler_syntax_end.length) === "\n```" &&
+        text.slice(range.start - 1, range.start) === "\n" &&
+        text.includes("```spoiler ") &&
+        text.lastIndexOf("```spoiler ") < range.start - 11;
+
+    if (is_curser_at_header_position()) {
+        const new_range_end = text.indexOf("\n```", range.start);
+        text =
+            text.slice(0, text.indexOf(spoiler_syntax_start)) +
+            text.slice(range.start + 1, new_range_end) +
+            text.slice(new_range_end + 4, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start - 11, new_range_end);
+        return;
+    } else if (is_inner_text_formatted("```spoiler", spoiler_syntax_end, range, selected_text)) {
+        text =
+            text.slice(0, range.start) +
+            text.slice(range.start + 11, range.end - 4) +
+            text.slice(range.end, text.length);
+        if (text.startsWith("\n")) {
+            text = text.slice(1, text.length);
+        }
+        set(field, text);
+        field.setSelectionRange(range.start, range.end);
+        return;
+    } else if (is_content_selected()) {
+        text =
+            text.slice(0, text.indexOf("```spoiler")) +
+            selected_text +
+            text.slice(range.end + spoiler_syntax_end.length, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start, range.end);
+        return;
+    } else if (is_selection_formatted("```spoiler ", spoiler_syntax_end, range, text)) {
+        text =
+            text.slice(0, text.indexOf("```spoiler")) +
+            selected_text +
+            text.slice(range.end + 4, text.length);
+        set(field, text);
+        field.setSelectionRange(range.start, range.end);
```

Comment:
Is `setSelectionRange` the right function to use for this sort of manipulation?  I worry that it won't work with `Ctrl+Z` correctly.